lib/plugin.js

```diff
@@ -226,28 +232,58 @@ internals.implementation = function (server, options) {
 };
 
 
+const getTokenFromCookie = (cookieHeader, cookieName) => {
+
+    if (!(cookieHeader || cookieName)) {
+        return null;
+    }
+
+    const cookies = cookieHeader.split(/;(\s+)?/);
+
+    // find the cookie with the cookie name given in the settings
+    for (let i = 0; i < cookies.length; i += 1) {
+        const [name, value] = cookies[i].split('=');
+
+        if (name === cookieName) {
+            return value;
+        }
+    }
+
+    return '';
+};
+
+
 internals.token = function (request, settings, missing, unauthorized) {
 
-    const authorization = request.headers.authorization;
+    // if both the headerName and cookieName are given false then we should fallback to
+    // authorization header as the default
+    const authorization =
+        request.headers[settings.headerName]
+            || getTokenFromCookie(request.headers.cookie, settings.cookieName)
+            || request.headers.authorization;
+
     if (!authorization) {
         throw missing;
     }
 
-    // Parse authorization header
+    // authorization header will be like <Scheme> <Token>
+    if (settings.headerName) {
+        const parts = authorization.split(/\s+/);
+        if (parts[0].toLowerCase() !== settings.httpAuthScheme.toLowerCase()) {
+            throw missing;
+        }
 
-    const parts = authorization.split(/\s+/);
-    if (parts[0].toLowerCase() !== settings.httpAuthScheme.toLowerCase()) {
-        throw missing;
-    }
+        if (parts.length !== 2) {
+            throw unauthorized('Bad HTTP authentication header format');
+        }
 
-    if (parts.length !== 2) {
-        throw unauthorized('Bad HTTP authentication header format');
-    }
+        const token = parts[1];
+        if (!token) {
+            throw missing;
+        }
 
-    const token = parts[1];
-    if (!token) {
-        throw missing;
+        return token;
```

Comment:
If both `headerName` and `cookieName` is specified, this logic will apply to the cookie value when the `headerName` header is not available.

Reply:
Will it ? Because when I tried to put cookie value as "Bearer <token value here>" hapi rejected the token saying it's invalid. Also According to what I read space is not allowed in cookies as quoted from [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#attributes). So checking for http auth scheme is not needed as the scheme would not be passed, right ?

Reply:
The logic will still trigger on `cookie: token=whatever`. It's true that it will effectively do nothing for all valid inputs, but it still shouldn't run.

Reply:
Also, this should be backed by a test case.